### PR TITLE
enhance: strip out theming on copy and paste from preview

### PR DESCRIPTION
**Pull request by nickolay-kondratyev**
# Pull Request Checklist

Note as mentioned in the comments using the clipboard readText() already strips out the theming. However, that means the headers are written as plain text. 

### Quality Assurance
- [x] Add a [test](https://wiki.dendron.so/notes/cb22bd36-d45a-4ffd-a31e-96c4b39cb7fb.html#writing-tests) for the new feature
- [x] Make sure all the existing [tests](https://wiki.dendron.so/notes/cb22bd36-d45a-4ffd-a31e-96c4b39cb7fb.html#running-all-tests) pass
- [x] Do a spot check by running your feature with our [test workspace](https://wiki.dendron.so/notes/cb22bd36-d45a-4ffd-a31e-96c4b39cb7fb.html#test-workspace)
- [x] If your tests involve running [[manual Testing|dendron://dendron.dendron-site/dendron.dev.qa.test#manual-testing]], make sure to update [[Test Workspace|dendron://dendron.dendron-site/dendron.dev.ref.test-workspace]] with any necessary notes/additions to perform manual test ^9jtWc6ov340p
- [ ] After you submit your pull request, check the output of our [integration test](https://github.com/dendronhq/dendron/actions) and make sure all tests pass
  - NOTE: If you running MacOS or Linux, pay special attention to the Windows output and vice versa if you are developing on Windows

### CODE REVIEW

File: packages/common-server/src/etc.ts [hunk 1]
```diff
@@ -61,7 +61,13 @@ export class WebViewCommonUtils {
 
     <script type="text/javascript">
       var theme = 'unknown';
-
+      
+      document.addEventListener('copy', (e) => {
```
Comment (kevinslin) [inline packages/common-server/src/etc.ts:65]: can you fix the formatting here to be in line with the rest of the doc?

**Review by kevinslin — CHANGES_REQUESTED** [review]
the main benefit of copying from preview is to preserve HTML attributes (eg. c/p to gmail and preserve links). 

can we modify the formatting as in [Remove background color from code copied from Visual Studio Code](https://gist.github.com/pragdave/6fc95ff2420dea51842717a5da7a98ff)?

some more details on modifying copy data from javascript here -> [Modify Clipboard content after copy event: JavaScript, jQuery](https://stackoverflow.com/questions/42089713/modify-clipboard-content-after-copy-event-javascript-jquery)

File: packages/common-server/src/etc.ts (hunk 1 above)
Comment (nickolay-kondratyev) [reply packages/common-server/src/etc.ts:65]: Will need to circle back to this one at another time thus far not going far.

# Some further attempts
Copying preview of:
```
## XVault Link

- vault1 bar: [[dendron://vault/bar]]
- vault2 bar: [[dendron://vault2/bar]]

## URL
[[https://wiki.dendron.so/#getting-started]]
```

## In JS
In JS with the following code:
```
      document.addEventListener('copy', (e) => {

          console.log("--------------------------------------------------------------------------------");
          console.log("---window.getSelection().toString()---");
          console.log(window.getSelection().toString());

          console.log("---JSON.stringify(window.getSelection())---");
          console.log(JSON.stringify(window.getSelection()));

          console.log("---e.clipboardData.getData('text/html')---");
          console.log(e.clipboardData.getData('text/html'));

          console.log("---window.clipboard.readText()---");
          if (window.clipboard === undefined){
               console.log("window.clipboard is UNDEFINED");
          }else{
              console.log(window.clipboard.readText());
          }
          console.log("--------------------------------------------------------------------------------");

          window.vscode.postMessage({
              type: 'onCopy',
          });
      });
```


```
--------------------------------------------------------------------------------
---window.getSelection().toString()---
XVault Link
vault1 bar: bar
vault2 bar: Bar
URL
https://wiki.dendron.so/

---JSON.stringify(window.getSelection())---
{}

---e.clipboardData.getData('text/html')---

---window.clipboard.readText()---
window.clipboard is UNDEFINED
--------------------------------------------------------------------------------
```

## Other APIs

Looking around the APIs this API sounds like it should return more than plain text:
https://developer.mozilla.org/en-US/docs/Web/API/Clipboard/read

However, VSCode interface for clipboard appears only to have read text for read:
```
    export interface Clipboard {

        /**
         * Read the current clipboard contents as text.
         * @returns A thenable that resolves to a string.
         */
        readText(): Thenable<string>;

        /**
         * Writes text into the clipboard.
         * @returns A thenable that resolves when writing happened.
         */
        writeText(value: string): Thenable<void>;
}
```

## Trying navigator API

```

      document.addEventListener('copy', async (e) => {

          console.log("COPY EVENT--------------------------------------------------------------------------------");
          const clipboardItems = await window.navigator.clipboard.read();

          console.log(clipboardItems.length);
          console.log(clipboardItems);
          console.log(JSON.stringify(clipboardItems));

          console.log("--------------------------------------------------------------------------------");

          window.vscode.postMessage({
              type: 'onCopy',
          });
      });

        document.addEventListener('paste', async (e) => {

          console.log("PASTE EVENT--------------------------------------------------------------------------------");
          const clipboardItems = await window.navigator.clipboard.read();
          console.log(clipboardItems.length);
          console.log(clipboardItems);
          console.log(JSON.stringify(clipboardItems));

          console.log("--------------------------------------------------------------------------------");


      });
```

Output
```
COPY EVENT--------------------------------------------------------------------------------
1
[ClipboardItem]
[{}]
--------------------------------------------------------------------------------
PASTE EVENT--------------------------------------------------------------------------------
1
[ClipboardItem]
[{}]
--------------------------------------------------------------------------------
```
When browsing the content of Clipboard item not able to see the data that was copied.

Comment (kevinslin) [reply packages/common-server/src/etc.ts:65]: 1. can you set a `debugger` in the webview to see if there are other properties of the event that have the output? see [[Qa|dendron://dendron.docs/pkg.dendron-plugin-views.qa]]

2. there might be more information in the following places
- [Clipboard VSCOde Issues · microsoft/vscode](https://github.com/microsoft/vscode/labels/editor-clipboard): find clipboard issues might provide more clues
    - [Add setting to keep tab characters in HTML clipboard · Issue #134588 · microsoft/vscode](https://github.com/microsoft/vscode/issues/134588)
    - [Expose Clipboard#read/write with data transfer · Issue #77790 · microsoft/vscode](https://github.com/microsoft/vscode/issues/77790)
- [vscode-paste-special/extension.ts at master · d3v-one/vscode-paste-special](https://github.com/d3v-one/vscode-paste-special/blob/master/src/extension.ts): extension that uses native commands to update clipboard settings

3. if you don't find an adequate solution, can you submit an issue to vscode?

Comment (nickolay-kondratyev) [reply packages/common-server/src/etc.ts]: In the new commit: tried to approach it by taking the selection on copy event and taking just the HTML.

Website Links appear to work as links and headers appear as headers (when header and something is copied). 

The background is removed.

WikiLinks to note as expected will have nothing to link. They are copied as what appears to be an empty link (retains correct name of the link though).

**Comment by kevinslin** [comment]
looks like there are some conflicts

**Comment by nickolay-kondratyev** [comment]
Conflicts resolved.

File: packages/common-server/src/etc.ts [hunk 2]
```diff
@@ -63,16 +63,16 @@ export class WebViewCommonUtils {
...
-          applyTheme(document.body.className);
-      
-          var observer = new MutationObserver(function(mutations) {
-              mutations.forEach(function(mutationRecord) {
-                  applyTheme(mutationRecord.target.className);
-              });    
-          });
-          var target = document.body;
-          observer.observe(target, { attributes : true, attributeFilter : ['class'] });
+        console.log("calling onLoad");
```
Comment (kevinslin) [inline packages/common-server/src/etc.ts:66]: remove console.log

**Review by kevinslin — APPROVED** [review]
just one minor comment about console.log. feel free to merge afterwards (and add to weekly ready to ship)

**Pull request merged**